src/OtcOption.sol

```diff
@@ -0,0 +1,252 @@
+
+error Option__InvalidUnderlyingToken();
+error Option__InvalidQuoteToken();
+error Option__InvalidStrike();
+error Option__InvalidMaturity();
+error Option__InvalidAmount();
+error Option__InvalidPremium();
+
+error Option__PriceFeedNotFound();
+error Option__DepositPremiumFailed();
+error Option__DepositMarginFailed();
+error Option__TransferPremiumFailed();
+error Option__TransferMarginFailed();
+error Option__DealNotOpen();
+error Option__DealExpired();
+error Option__DealNotExpired();
+error Option__DealNotFound();
+error Option__DealHasNoTaker();
+
+contract OtcOption is Ownable {
+    event DealCreated(
+        uint id,
+        address underlyingToken,
+        address quoteToken,
+        uint strike,
+        uint maturity,
+        bool isCall,
+        uint amount,
+        uint premium,
+        bool isMakerBuyer,
+        address maker
+    );
+    event DealTaken(uint id, address taker);
+    event DealRemoved(uint id);
+    event DealSettled(uint id, bool exercised);
+
+    enum DealStatus {
+        Open,
+        Taken,
+        Removed,
+        Settled
+    }
+
+    struct Deal {
+        uint id;
+        address underlyingToken;
+        address quoteToken;
+        uint strike;
+        uint maturity;
+        bool isCall;
+        uint amount;
+        uint premium;
+        bool isMakerBuyer;
+        DealStatus status;
+        address maker;
+        address taker;
+    }
+
+    // status variables
+    mapping(address => mapping(address => address)) private _priceFeeds; // underlying => (quote => feed)
+    uint private _dealCounter = 1;
+    mapping(uint => Deal) private _deals;
+
+    function addPriceFeed(
+        address _underlyingToken,
+        address _quoteToken,
+        address _priceFeedAddress
+    ) public onlyOwner {
+        _priceFeeds[_underlyingToken][_quoteToken] = _priceFeedAddress;
+    }
+
+    function getPriceFeed(
+        address _underlyingToken,
+        address _quoteToken
+    ) public view returns (address) {
+        address priceFeed = _priceFeeds[_underlyingToken][_quoteToken];
+        if (priceFeed == address(0)) revert Option__PriceFeedNotFound();
+        return priceFeed;
+    }
+
+    function createDeal(
+        address _underlyingToken,
+        address _quoteToken,
+        uint _strike,
+        uint _maturity,
+        bool _isCall,
+        uint _amount,
+        uint _premium,
+        bool _isMakerBuyer
+    ) external returns (uint) {
+        if (_underlyingToken == address(0)) revert Option__InvalidUnderlyingToken();
+        if (_quoteToken == address(0)) revert Option__InvalidQuoteToken();
+        if (_strike == 0) revert Option__InvalidStrike();
+        if (_maturity <= block.timestamp) revert Option__InvalidMaturity();
+        if (_amount == 0) revert Option__InvalidAmount();
+        if (_premium == 0) revert Option__InvalidPremium();
+
+        Deal memory deal = Deal(
+            _dealCounter,
```

Comment:
Since _dealCounter grows linearly for each call, we can `unchecked {...}` reducing gas and put here ++_dealCounter